### PR TITLE
Fix and complete dwmapi; add uxtheme

### PR DESCRIPTION
**Pull request by MSxDOS**

### CODE REVIEW

File: src/um/uxtheme.rs [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+    ) -> HRESULT;
+}
+#[inline]
+pub unsafe fn SetWindowThemeNonClientAttributes(
+    hwnd: HWND,
+    dwMask: DWORD,
+    dwAttributes: DWORD,
+) -> HRESULT {
+    use core::mem::{size_of, uninitialized};
+    let mut wta: WTA_OPTIONS = uninitialized();
```
Comment (retep998) [inline src/um/uxtheme.rs]: Can you use `zeroed` here instead?

File: src/um/uxtheme.rs [hunk 2]
```diff
@@ -0,0 +1,755 @@
...
+    pub fn EndBufferedPaint(
+        hBufferedPaint: HPAINTBUFFER,
+        fUpdateTarget: BOOL,
+    ) -> HRESULT;
+    pub fn GetBufferedPaintTargetRect(
+        hBufferedPaint: HPAINTBUFFER,
+        prc: *mut RECT,
+    ) -> HRESULT;
+    pub fn GetBufferedPaintTargetDC(hBufferedPaint: HPAINTBUFFER) -> HDC;
+    pub fn GetBufferedPaintDC(hBufferedPaint: HPAINTBUFFER) -> HDC;
```
Comment (retep998) [inline src/um/uxtheme.rs]: These should still be block indented!

File: src/um/uxtheme.rs [hunk 3]
```diff
@@ -0,0 +1,755 @@
...
+    pub fn BufferedPaintClear(
+        hBufferedPaint: HPAINTBUFFER,
+        prc: *const RECT,
+    ) -> HRESULT;
+    pub fn BufferedPaintSetAlpha(
+        hBufferedPaint: HPAINTBUFFER,
+        prc: *const RECT,
+        alpha: BYTE,
+    ) -> HRESULT;
+    pub fn BufferedPaintStopAllAnimations(hwnd: HWND) -> HRESULT;
```
Comment (retep998) [inline src/um/uxtheme.rs]: And this!

**Review by retep998 — CHANGES_REQUESTED** [review]

File: src/um/dwmapi.rs [hunk 4]
```diff
@@ -63,32 +118,51 @@ STRUCT!{#[repr(packed)] struct DWM_TIMING_INFO {
...
+    eSampling: DWM_SOURCE_FRAME_SAMPLING,
+}}
+// pub const DWM_FRAME_DURATION_DEFAULT: i32 = -1;
+FN!{stdcall DwmDefWindowProc(
+        hWnd: HWND,
+        msg: UINT,
+        wParam: WPARAM,
+        lParam: LPARAM,
+        plResult: *mut LRESULT,
+    ) -> BOOL}
```
Comment (retep998) [inline src/um/dwmapi.rs]: What's with the indentation here?

File: src/um/dwmapi.rs [hunk 5]
```diff
@@ -99,31 +173,114 @@ extern "system" {
         pvAttribute: LPVOID,
         cbAttribute: DWORD,
     ) -> HRESULT;
-    // pub fn DwmIsCompositionEnabled();
-    // pub fn DwmModifyPreviousDxFrameDuration();
-    // pub fn DwmQueryThumbnailSourceSize();
-    // pub fn DwmRegisterThumbnail();
-    // pub fn DwmSetDxFrameDuration();
-    // pub fn DwmSetPresentParameters();
+    pub fn DwmIsCompositionEnabled(pfEnabled: *mut BOOL) -> HRESULT;
```
Comment (retep998) [inline src/um/dwmapi.rs]: More block indentation needed!

File: src/um/dwmapi.rs [hunk 6]
```diff
@@ -63,32 +118,51 @@ STRUCT!{#[repr(packed)] struct DWM_TIMING_INFO {
...
-    // pub fn DwmEnableComposition();
-    // pub fn DwmEnableMMCSS();
-    // pub fn DwmExtendFrameIntoClientArea();
-    // pub fn DwmGetColorizationColor();
+}
+pub const DWM_EC_DISABLECOMPOSITION: UINT = 0;
+pub const DWM_EC_ENABLECOMPOSITION: UINT = 1;
+extern "system" {
+    pub fn DwmEnableComposition(uCompositionAction: UINT) -> HRESULT;
+    pub fn DwmEnableMMCSS(fEnableMMCSS: BOOL) -> HRESULT;
```
Comment (retep998) [inline src/um/dwmapi.rs]: *cough*

File: src/um/dwmapi.rs [hunk 7]
```diff
@@ -99,31 +173,114 @@ extern "system" {
...
     ) -> HRESULT;
-    // pub fn DwmUnregisterThumbnail();
-    // pub fn DwmUpdateThumbnailProperties();
-    // pub fn DwmSetIconicThumbnail();
-    // pub fn DwmSetIconicLivePreviewBitmap();
-    // pub fn DwmInvalidateIconicBitmaps();
-    // pub fn DwmAttachMilContent();
-    // pub fn DwmDetachMilContent();
-    pub fn DwmFlush();
+    pub fn DwmUnregisterThumbnail(hThumbnailId: HTHUMBNAIL) -> HRESULT;
```
Comment (retep998) [inline src/um/dwmapi.rs]: *cough*

File: src/um/dwmapi.rs [hunk 8]
```diff
@@ -99,31 +173,114 @@ extern "system" {
...
+        hbmp: HBITMAP,
+        dwSITFlags: DWORD,
+    ) -> HRESULT;
+    pub fn DwmSetIconicLivePreviewBitmap(
+        hwnd: HWND,
+        hbmp: HBITMAP,
+        pptClient: *mut POINT,
+        dwSITFlags: DWORD,
+    ) -> HRESULT;
+    pub fn DwmInvalidateIconicBitmaps(hwnd: HWND) -> HRESULT;
```
Comment (retep998) [inline src/um/dwmapi.rs]: ***cough***

File: src/um/uxtheme.rs [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+use shared::minwindef::{BOOL, BYTE, DWORD, HINSTANCE, HRGN, LPARAM, UINT, ULONG, WORD};
+use shared::windef::{COLORREF, HBITMAP, HBRUSH, HDC, HWND, LPCRECT, LPRECT, POINT, RECT, SIZE};
+use um::commctrl::HIMAGELIST;
+use um::wingdi::{BLENDFUNCTION, LOGFONTW, RGBQUAD, TEXTMETRICW};
+use um::winnt::{HANDLE, HRESULT, LONG, LPCWSTR, LPWSTR, PVOID, VOID};
+pub type HTHEME = HANDLE;
+//pub const MAX_THEMECOLOR: u32 = 64;
+//pub const MAX_THEMESIZE: u32 = 64;
+extern "system" {
+    pub fn BeginPanningFeedback(hwnd: HWND) -> BOOL;
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***COUGH***

File: src/um/uxtheme.rs [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+        hwnd: HWND,
+        pszClassList: LPCWSTR,
+        dpi: UINT,
+    ) -> HTHEME;
+    pub fn OpenThemeDataEx(
+        hwnd: HWND,
+        pszClassList: LPCWSTR,
+        dwFlags: DWORD,
+    ) -> HTHEME;
+    pub fn CloseThemeData(hTheme: HTHEME) -> HRESULT;
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***C̦̹̠̭͇͔͂̃̂̄̚̕͞O̫͈͍͖̠̣̻̊͒̀̾̔̎̊Ự̶͚͇̜̰͕͎͉͎́̍̇̀͛G̵͚̳͍̫͖̙͉̊͛̀͢͠͡Ḩ̹͔͈͖͔͔͗̄̐̍͜͝͝***

File: src/um/uxtheme.rs [hunk 11]
```diff
@@ -0,0 +1,761 @@
...
+        cchMaxStringChars: c_int,
+    ) -> HRESULT;
+    pub fn GetThemeSysInt(
+        hTheme: HTHEME,
+        iIntId: c_int,
+        piValue: *mut c_int,
+    ) -> HRESULT;
+    pub fn IsThemeActive() -> BOOL;
+    pub fn IsAppThemed() -> BOOL;
+    pub fn GetWindowTheme(hwnd: HWND) -> HTHEME;
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***C̶̢̧̗̗͕̮͚͆̾̊͌͊̒̍̄͠Ơ̴̲͇͉̪͉̳̯͔̍̊̌̏̉̋̀̕͜͡U͎̳̻͎̠̮̜̜͑̒̿͒́͘͘͠͡G̨̡̹̥͎̱̓̒͋͆̽́̊̋͟͜͜H̷͓͕̣̟̲̋̊̿̐̌͠***

File: src/um/uxtheme.rs [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+pub const ETDT_USEAEROWIZARDTABTEXTURE: DWORD = 0x00000008;
+pub const ETDT_ENABLEAEROWIZARDTAB: DWORD = ETDT_ENABLE | ETDT_USEAEROWIZARDTABTEXTURE;
+pub const ETDT_VALIDBITS: DWORD = ETDT_DISABLE | ETDT_ENABLE | ETDT_USETABTEXTURE
+    | ETDT_USEAEROWIZARDTABTEXTURE;
+extern "system" {
+    pub fn EnableThemeDialogTexture(
+        hwnd: HWND,
+        dwFlags: DWORD,
+    ) -> HRESULT;
+    pub fn IsThemeDialogTextureEnabled(hwnd: HWND) -> BOOL;
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***C̻̼̝̬̟̩͚̜̿̈́̽̉̿̅͆̚ͅÖ̧͔̮̪̫̟̱̰́͂̉̊͂̋͐͘͢͠͝ͅŲ̡̬̣̭̗̪̥̩̓̀͂̇̏̑͋͊̀̀Ģ̷̣̥͈͇͕͓̲͍͌̿̎̾͂̆͊H̴̙̙͈̙͇̼͙͐̇̊̃́ͅ***

File: src/um/uxtheme.rs [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+}
+pub const STAP_ALLOW_NONCLIENT: DWORD = 1 << 0;
+pub const STAP_ALLOW_CONTROLS: DWORD = 1 << 1;
+pub const STAP_ALLOW_WEBCONTENT: DWORD = 1 << 2;
+pub const STAP_VALIDBITS: DWORD = STAP_ALLOW_NONCLIENT | STAP_ALLOW_CONTROLS
+    | STAP_ALLOW_WEBCONTENT;
+extern "system" {
+    pub fn GetThemeAppProperties() -> DWORD;
+    pub fn SetThemeAppProperties(dwFlags: DWORD);
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***C̸̥͚̣̰͇̼̙͔̮̹̄̏̈̒̓̚O̸̰͕̘̹̫̮̲̜̍͗̎͛͛̇͡Ư͈̜͕͕͕̂̽̃̍̌͢͞͡͡G̷̜̞̜͕͓̪͔͕̦͑̍̋̋̚͠H̴̘͖̲̝̦͔̹̙̤̑̌̑̉́̐̋***

File: src/um/uxtheme.rs [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+        pszPropertyName: LPCWSTR,
+        pszValueBuff: LPWSTR,
+        cchMaxValChars: c_int,
+    ) -> HRESULT;
+    pub fn DrawThemeParentBackground(
+        hwnd: HWND,
+        hdc: HDC,
+        prc: *const RECT,
+    ) -> HRESULT;
+    pub fn EnableTheming(fEnable: BOOL) -> HRESULT;
```
Comment (retep998) [inline src/um/uxtheme.rs]: ***C̴͔͇͚͈͉̠̅̒̇̄́͐͆̆͟͠Ȍ̶̝͕̩̬̦̩̻̅͑̿̽̎̓Ǔ̸̢͍̦͔͎́̅͑͌͝ͅG̢͍̫̪̖̞̥̒̎̿͌̄͋̍̕͜͠H̵̰̞̹͙͖͔͇̪͇͕͋͌̀̉̈́͂̈́̀***

**Review by retep998 — CHANGES_REQUESTED** [review]

**Comment by MSxDOS** [comment]
Using existing bindings as an example is obviously a bad idea with `winapi`

File: src/um/dwmapi.rs [hunk 15]
```diff
@@ -63,32 +118,55 @@ STRUCT!{#[repr(packed)] struct DWM_TIMING_INFO {
...
+    cRefreshesPerFrame: UINT,
+    eSampling: DWM_SOURCE_FRAME_SAMPLING,
+}}
+// pub const DWM_FRAME_DURATION_DEFAULT: i32 = -1;
+FN!{stdcall DwmDefWindowProc(
+        hWnd: HWND,
+        msg: UINT,
+        wParam: WPARAM,
+        lParam: LPARAM,
+        plResult: *mut LRESULT,
```
Comment (retep998) [inline src/um/dwmapi.rs]: These parameters are still all excessively indented.

**Review by retep998 — CHANGES_REQUESTED** [review]

File: src/um/dwmapi.rs (hunk 15 above)
Comment (MSxDOS) [reply src/um/dwmapi.rs]: Issues with indentations of everything related to macros in the IDE that I use + no help from rustfmt + my own blindness :smile:

**Review by retep998 — APPROVED** [review]

**Pull request merged**